src/interrupts.rs

```diff
@@ -0,0 +1,41 @@
+use crate::*;
+
+// NOTE: ESP-IDF-specific
+const SPINLOCK_FREE_INIT: u32 = 0xB33FFFFF;
+pub type EspCriticalSection = portMUX_TYPE;
+
+impl EspCriticalSection {
+    pub fn new() -> Self {
+        Self {
+            owner: SPINLOCK_FREE_INIT,
+            count: 0,
+        }
+    }
+
+    pub fn enter_critical(&mut self) {
```

Comment:
Unless I'm missing something, the current implementation of `CriticalSection` cannot be used for an implementation of a "mutex". (Is this why you removed the mutex implementation?). There are two problems with it:
* `CriticalSection` does not implement `Send` and `Sync`, and as such, cannot be shared across multiple threads (say, by wrapping it in an `Arc`).
* But even if it did implement `Send` and `Sync`, the fact that `enter_critical` and `exit_critical` take a `&mut self` make it basically impossible to wrap in an `Arc`, as `Arc`s can only provide shared references, and do not provide mutable references.
* To me, in its current shape, `CriticalSection` allows only to disable interrupts for the CPU which happens to be the active one when `enter_critical`. The only mutex implementation where `CriticalSection` can be used would be of a one that is shared across threads using static mut variables, which are unfortunately unsafe themselves (besides being not so nice to use due to the fact that they are, well, static). This reminds of the limited usability of the [embedded mutex trait](https://github.com/rust-embedded/mutex-trait) which is - IMO - not very usable, and which is why we are having [our own mutex trait in embedded-svc](https://github.com/esp-rs/embedded-svc/blob/master/src/mutex.rs#L6)

Since I'm anyway in the midst of merging this code into `esp-idf-hal`, I just want to get your feedback whether I'm missing something or whether my logic is incorrect? If not, then my plan is to change `enter_critical` / `exit_critical` to take a shrared `&self` reference by wrapping the `portMUX_TYPE` in an `UnsafeCell` instance. Using `UnsafeCell` in this context should be safe (pun not intended) because `UnsafeCell::get()` (which is an unsafe call returning a `*mut PORTMUX_TYPE`) will only be used to send this reference to `vPortEnterCritical`/`vPortReleaseCritical` and the content of the `UnsafeCell` instance will not be otherwise shared with the rest of the Rust program.

Reply:
One more *significant* complication I should've considered from the very beginning: disabling interrupts is actually very dangerous on the esp32* chips, as you have to first of all make sure that **all code** that you execute from the moment interrupts are disabled till the moment they are re-enabled again is in IRAM memory. This is done (in C) by placing an `__IRAM` attribute in front of your functions, which can be verbose and error prone, or by using special esp32 linker scripts. 

Similarly - for Rust code - you should put (I think) a `#[link_section = ".rwtext"]` attribute in front of each and every function and method that might happen to be called when interrupts are disabled. Not sure how linker scripts would work for Rust code though. The reason for this is that when interrupts are disabled, the CPU cannot load code sections from flash into IRAM, so the user has to ensure that everything which is to be executed is already in IRAM.

Considering the above, I'm still willing to merge your `CriticalSection` code (with the changes I made so that interrupt-based mutexes are supported, as well as a few other utilities), however these will be hidden behind the `--features experimental` flag, until we make sure that the whole thing with the `__IRAM` attribute really works reliably.

Reply:
> One more significant complication I should've considered from the very beginning: disabling interrupts is actually very dangerous on the esp32* chips, as you have to first of all make sure that all code that you execute from the moment interrupts are disabled till the moment they are re-enabled again is in IRAM memory. This is done (in C) by placing an __IRAM attribute in front of your functions, which can be verbose and error prone, or by using special esp32 linker scripts.

@ivmarkov  I don't think this is the case. There is a similar looking issue, that interrupt handlers should be placed into IRAM if you want them to allow preempting flash operations. However simply disabling interrupts doesn't harm execution of code from Flash.

Reply:
@igrr Reading the ESP32 Technical Reference MMU section did not help me answer this simple question: is it the case, that - once configured - the MMU does NOT need any assistance from the CPU whatsoever in dealing with reads from flash? I.e. are you saying that I can have all interrupts disabled on the CPU, yet, the MMU will operate perfectly fine, handling cache misses, readings from flash and so on?

Reply:
That's correct, Cache/MMU hardware works with external Flash or PSRAM without relying on CPU. You can disable the interrupts and execution of code from Flash (or PSRAM, on chips which support this) will still work fine.